File: sandbox/php/test/SignatureRequestTest.php
<?php
declare(strict_types=1);

namespace Dropbox\SignSandbox\Tests;

use Dropbox\Sign;
use GuzzleHttp\Cookie\CookieJar;
use PHPUnit\Framework\TestCase;
use SplFileObject;

/**
 * This test suite is intended solely as a stopgap while we setup automated
 * internal tests from github actions.
 *
 * For now it requires running manually
 */
class SignatureRequestTest extends TestCase
{
    public const FIXTURES_DIR = __DIR__ . '/../test_fixtures';

    private Sign\Configuration $config;

    private string $client_id;

    public function setUp(): void
    {
        parent::setUp();

        $config_custom = json_decode(file_get_contents(__DIR__ . '/.config.json'), true);
        $config_dist = json_decode(file_get_contents(__DIR__ . '/.config.dist.json'), true);
        $config = array_merge($config_dist, $config_custom);

        $this->client_id = $config['CLIENT_ID'];

        $this->config = new Sign\Configuration();
        $this->config->setUsername($config['API_KEY']);
        $this->config->setHost($config['BASE_URL']);

        if ($config['USE_XDEBUG']) {
            $cookies = CookieJar::fromArray(
                ['XDEBUG_SESSION' => 'xdebug'],
                parse_url($config['BASE_URL'], PHP_URL_HOST),
            );

            $this->config->setOptions(['cookies' => $cookies]);
        }
    }

    public function testSend(): void
    {
        $signature_request_api = new Sign\Api\SignatureRequestApi($this->config);

        $data = json_decode(
            file_get_contents(self::FIXTURES_DIR . '/SignatureRequestSendRequest.json'),
            true,
        );
        $file = new SplFileObject(self::FIXTURES_DIR . '/pdf-sample.pdf');
        $data['files'] = [$file];

        $send_request = Sign\Model\SignatureRequestSendRequest::init($data);

        $send_response = $signature_request_api->signatureRequestSend($send_request);

        $this->assertEquals(
            $send_request->getFormFieldsPerDocument()[0]->getApiId(),
            $send_response->getSignatureRequest()->getCustomFields()[0]->getApiId(),
        );

        $this->assertEquals(
            $send_request->getSigners()[0]->getEmailAddress(),
            $send_response->getSignatureRequest()->getSignatures()[0]->getSignerEmailAddress(),
        );
        $this->assertEquals(
            $send_request->getSigners()[1]->getEmailAddress(),
            $send_response->getSignatureRequest()->getSignatures()[1]->getSignerEmailAddress(),
        );
        $this->assertEquals(
            $send_request->getSigners()[2]->getEmailAddress(),
            $send_response->getSignatureRequest()->getSignatures()[2]->getSignerEmailAddress(),
        );

        $get_response = $signature_request_api->signatureRequestGet(
            $send_response->getSignatureRequest()->getSignatureRequestId(),
        );

        $this->assertSame(
            $send_response->getSignatureRequest()->getSignatureRequestId(),
            $get_response->getSignatureRequest()->getSignatureRequestId(),
        );
    }

    public function testCreateEmbedded(): void
    {
        $signature_request_api = new Sign\Api\SignatureRequestApi($this->config);

        $data = json_decode(
            file_get_contents(self::FIXTURES_DIR . '/SignatureRequestCreateEmbeddedRequest.json'),
            true,
        );
        $file = new SplFileObject(self::FIXTURES_DIR . '/pdf-sample.pdf');
        $data['files'] = [$file];
        $data['client_id'] = $this->client_id;

        $send_request = Sign\Model\SignatureRequestCreateEmbeddedRequest::init($data);

        $send_response = $signature_request_api->signatureRequestCreateEmbedded($send_request);

        $this->assertEquals(
            $send_request->getSigners()[0]->getEmailAddress(),
            $send_response->getSignatureRequest()->getSignatures()[0]->getSignerEmailAddress(),
        );
        $this->assertEquals(
            $send_request->getSigners()[1]->getEmailAddress(),
            $send_response->getSignatureRequest()->getSignatures()[1]->getSignerEmailAddress(),
        );
        $this->assertEquals(
            $send_request->getSigners()[2]->getEmailAddress(),
            $send_response->getSignatureRequest()->getSignatures()[2]->getSignerEmailAddress(),
        );

        $embedded_api = new Sign\Api\EmbeddedApi($this->config);

        $get_response = $embedded_api->embeddedSignUrl(
            $send_response->getSignatureRequest()->getSignatures()[0]->getSignatureId(),
        );

        $this->assertNotEmpty($get_response->getEmbedded()->getSignUrl());
    }

    public function testSendWithoutFileError(): void
    {
        $signature_request_api = new Sign\Api\SignatureRequestApi($this->config);

        $data = json_decode(
            file_get_contents(self::FIXTURES_DIR . '/SignatureRequestSendRequest.json'),
            true,
        );

        $request = Sign\Model\SignatureRequestSendRequest::init($data);

        try {
            $response = $signature_request_api->signatureRequestSend($request);

            $this->fail('Should have thrown: ' . print_r($response, true));
        } catch (Sign\ApiException $e) {
            $error = $e->getResponseObject();

            $this->assertEquals('file', $error->getError()->getErrorPath());
        }
    }
}

File: sandbox/java-v2/src/test/com/dropbox/sign_sandbox/SignatureRequestTest.java
package com.dropbox.sign_sandbox;

import com.dropbox.sign.ApiClient;
import com.dropbox.sign.ApiException;
import com.dropbox.sign.Configuration;
import com.dropbox.sign.JSON;
import com.dropbox.sign.api.EmbeddedApi;
import com.dropbox.sign.api.SignatureRequestApi;
import com.dropbox.sign.model.*;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectReader;
import org.junit.Assert;
import org.junit.Test;

import java.io.File;
import java.nio.file.Files;
import java.nio.file.Paths;

public class SignatureRequestTest {
    public static JsonNode getConfig() throws Exception {
        ObjectMapper objectMapper = JSON.getDefault().getMapper();

        JsonNode configCustom = objectMapper.readTree(Files.newInputStream(
            Paths.get("src/test/com/dropbox/sign_sandbox/.config.json")
        ));
        JsonNode configDist = objectMapper.readTree(Files.newInputStream(
            Paths.get("src/test/com/dropbox/sign_sandbox/.config.dist.json")
        ));

        ObjectReader updater = objectMapper.readerForUpdating(configDist);

        return updater.readValue(configCustom);
    }

    public static JsonNode getJsonContents(String filename) throws Exception {
        ObjectMapper objectMapper = JSON.getDefault().getMapper();
        return objectMapper.readTree(Files.newInputStream(
            Paths.get("test_fixtures/" + filename)
        ));
    }

    @Test
    public void testSend() throws Exception {
        JsonNode config = getConfig();

        ApiClient apiClient = Configuration.getDefaultApiClient()
            .setApiKey(config.findPath("API_KEY").textValue())
            .setBasePath(config.findPath("BASE_URL").textValue());

        SignatureRequestApi signatureRequestApi = new SignatureRequestApi(apiClient);

        JsonNode data = getJsonContents("SignatureRequestSendRequest.json");

        SignatureRequestSendRequest sendRequest = SignatureRequestSendRequest.init(data.toString());
        sendRequest.addFilesItem(new File("test_fixtures/pdf-sample.pdf"));

        SignatureRequestGetResponse sendResponse = signatureRequestApi.signatureRequestSend(sendRequest);

        Assert.assertEquals(
            sendRequest.getFormFieldsPerDocument().get(0).getApiId(),
            sendResponse.getSignatureRequest().getCustomFields().get(0).getApiId()
        );

        Assert.assertEquals(
            sendRequest.getSigners().get(0).getEmailAddress(),
            sendResponse.getSignatureRequest().getSignatures().get(0).getSignerEmailAddress()
        );
        Assert.assertEquals(
            sendRequest.getSigners().get(1).getEmailAddress(),
            sendResponse.getSignatureRequest().getSignatures().get(1).getSignerEmailAddress()
        );
        Assert.assertEquals(
            sendRequest.getSigners().get(2).getEmailAddress(),
            sendResponse.getSignatureRequest().getSignatures().get(2).getSignerEmailAddress()
        );

        SignatureRequestGetResponse getResponse = signatureRequestApi.signatureRequestGet(
            sendResponse.getSignatureRequest().getSignatureRequestId()
        );

        Assert.assertEquals(
            sendResponse.getSignatureRequest().getSignatureRequestId(),
            getResponse.getSignatureRequest().getSignatureRequestId()
        );
    }

    @Test
    public void testCreateEmbedded() throws Exception {
        JsonNode config = getConfig();

        ApiClient apiClient = Configuration.getDefaultApiClient()
            .setApiKey(config.findPath("API_KEY").textValue())
            .setBasePath(config.findPath("BASE_URL").textValue());

        SignatureRequestApi signatureRequestApi = new SignatureRequestApi(apiClient);

        JsonNode data = getJsonContents("SignatureRequestCreateEmbeddedRequest.json");

        SignatureRequestCreateEmbeddedRequest sendRequest = SignatureRequestCreateEmbeddedRequest.init(data.toString());
        sendRequest.addFilesItem(new File("test_fixtures/pdf-sample.pdf"));
        sendRequest.clientId(config.findPath("CLIENT_ID").textValue());

        SignatureRequestGetResponse sendResponse = signatureRequestApi.signatureRequestCreateEmbedded(sendRequest);

        Assert.assertEquals(
            sendRequest.getSigners().get(0).getEmailAddress(),
            sendResponse.getSignatureRequest().getSignatures().get(0).getSignerEmailAddress()
        );
        Assert.assertEquals(
            sendRequest.getSigners().get(1).getEmailAddress(),
            sendResponse.getSignatureRequest().getSignatures().get(1).getSignerEmailAddress()
        );
        Assert.assertEquals(
            sendRequest.getSigners().get(2).getEmailAddress(),
            sendResponse.getSignatureRequest().getSignatures().get(2).getSignerEmailAddress()
        );

        EmbeddedApi embeddedApi = new EmbeddedApi(apiClient);

        EmbeddedSignUrlResponse getResponse = embeddedApi.embeddedSignUrl(
            sendResponse.getSignatureRequest().getSignatures().get(0).getSignatureId()
        );

        Assert.assertNotNull(getResponse.getEmbedded().getSignUrl());
    }

    @Test
    public void testSendWithoutFileError() throws Exception {
        JsonNode config = getConfig();

        ApiClient apiClient = Configuration.getDefaultApiClient()
            .setApiKey(config.findPath("API_KEY").textValue())
            .setBasePath(config.findPath("BASE_URL").textValue());

        SignatureRequestApi signatureRequestApi = new SignatureRequestApi(apiClient);

        JsonNode data = getJsonContents("SignatureRequestSendRequest.json");
        SignatureRequestSendRequest sendRequest = SignatureRequestSendRequest.init(data.toString());

        try {
            signatureRequestApi.signatureRequestSend(sendRequest);
            Assert.fail();
        } catch (ApiException e) {
            Assert.assertEquals("file", e.getErrorResponse().getError().getErrorPath());
        }
    }
}
